scripts/release/release.py

```diff
@@ -0,0 +1,103 @@
+import re
+
+from io import StringIO
+
+from helpers import run
+from tickets import check_tickets
+
+release_tag_pattern = re.compile(r"^v\d\.\d\.\d(-RC\d)?$")
+dry_run = False
+
+
+def get_latest_release_tag():
+    tags = run("git tag").split('\n')
+    release_tags = sorted(t for t in tags if release_tag_pattern.match(t))
+    return release_tags[-1]
+
+
+def git_is_clean():
+    return not run("git status -s")
+
+
+def tag(tag_name, branch_diff):
+    message = "Release {tag}, incorporating these branches: {branches}".format(
+        tag=tag_name, branches=branch_diff)
+    run("git tag -a {tag} -m \"{msg}\"".format(tag=tag_name, msg=message))
+
+
+def push():
+    if not dry_run:
+        run("git push --tags")
+
+
+def get_new_tag():
+    new_tag = "v" + input("What should the new release tag be? v")
+    if new_tag <= latest_tag:
+        template = "Error: {new_tag} is not after {latest_tag}"
+        print(template.format(new_tag=new_tag, latest_tag=latest_tag))
+        exit(-1)
+    if not release_tag_pattern.match(new_tag):
+        print("Error: tag does not correspond to regex")
+        exit(-1)
+    return new_tag
+
+
+def make_release_message(tag, branches_and_tickets):
+    print(branches_and_tickets)
+    with StringIO() as msg:
+        print("# " + tag, file=msg)
+        print("## Tickets", file=msg)
+        for branch, ticket in branches_and_tickets:
+            if ticket:
+                summary = ticket.get("summary")
+                line = "* {branch}: {summary}".format(branch=branch,
+                                                      summary=summary)
+                print(line, file=msg)
+
+        print("\n## Other branches merged in this release", file=msg)
+        for branch, ticket in branches_and_tickets:
+            if not ticket:
+                print("* " + branch, file=msg)
+        return msg.getvalue()
+
+
+def write_release_log(message):
+    with open('RELEASE_LOG.md', 'a') as f:
+        f.write(message)
+        f.write("\n")
+
+
+def commit_tag_and_push():
+    run("git add RELEASE_LOG.md")
+    run("git commit -m \"{msg}\"".format(msg=release_message))
+    print("Tagging and pushing...")
+    tag(new_tag, release_message)
+    push()
+
+
+def fetch():
+    print("Fetching from remote...")
+    run("git fetch --tags --all")
+
+
+if __name__ == "__main__":
+    if not (git_is_clean() or dry_run):
+        print("Git status reports as not clean; aborting release")
+    else:
+        fetch()
+        latest_tag = get_latest_release_tag()
+        print("The latest release was " + latest_tag)
+
+        branches_and_tickets = check_tickets(latest_tag)
+        new_tag = get_new_tag()
+
+        print("Writing release log...")
+        release_message = make_release_message(new_tag, branches_and_tickets)
+        write_release_log(release_message)
+        commit_tag_and_push()
+
+        print("Done!")
+        print("""The released tickets should now be moved into the 'Deployed'
```

Comment:
Technically this isn't true yet! This script will tag and push tag to git but not actually deploy the code... presumably the flow is run this on UAT, deploy to UAT, then post user acceptance testing deploy to prod? 

Could you update the README to explain the flow?

Reply:
Yeah, I wasn't sure about this. Maybe we want an extra state in YouTrack which is "Deployed to UAT". Then you run this script and it moves the tickets from "Ready to deploy" to "Deployed to UAT" (or, for now, prompts you to do that manually). Then we have another script that just moves the tickets from "Deployed to UAT" to "Deployed"?

Reply:
I think it only make sense for tickets to be automatically moved to 'deployed' (or 'deployed to UAT') by a script that does actually deploy code.

Really we want a pre-deploy step which is tagging the release (as done here), a deploy step (as we already have), and a post-deploy step, which is updating the tickets.

I think the flow we're aiming for ultimately is something like:

* deploy to UAT happens automatically on a successful Montagu build in Teamcity 
* post-deploy step on UAT moves tickets to 'Deployed to UAT' state
* manual user acceptance testing takes place
* release is created via manual intervention, i.e. running this script on UAT, which would then look at the tickets in 'deployed to UAT' state and include those in the release log
* deploy to prod (maybe `./src/deploy.py' should take a tag as command line argument here and automatically fetch that first)
* post-deploy step on prod moves tickets to 'Deployed' state

Reply:
Probably we should be thinking about external tools that can manage this flow. But either way I think for now just remove this prompt about tickets, and update the README so it's clear that this script creates a release, rather than deploying a release... 

Maybe the release log should actually have a flag to say whether a particular release was ever deployed (and when), set to false here and updated by the deploy script at the end of a successful deploy?

Reply:
Okay, for now I've just made it clear the script doesn't actually release anything. It's now called `make-release.py` and it doesn't tell you to update the tickets.